### PR TITLE
Spec: Rewrite the type system fundamentals.

### PR DESCRIPTION
**Pull request by sjrd**
There are two fundamental differences between the type sytems of Scala 2 and 3, which made a fundamental rewrite desirable:

- Scala 2 defines path-dependent types in terms of type projections, whereas Scala 3 is path-dependent first, and has not general type projects.
- Scala 2 has ad hoc higher-kinded type parameters and polymorphic type aliases, whereas Scala 3 has first-class type lambdas and is fundamentally higher kinded.

We therefore rewrite the entire type system section of the spec to be higher-kinded and path-dependent first. In the process, we also revise refined types to better fit the path-dependent nature.

I (sjrd) took the liberty to add myself as a contributor in the index page of the spec, as I believe this constitutes a fundamental contribution to the language specification.

---

The rendered version of the Types chapter can be seen here:
~~[Types Scala 3.pdf](https://github.com/lampepfl/dotty/files/11598665/Types.Scala.3.pdf)~~ (old version)
[Types Scala 3.pdf](https://github.com/lampepfl/dotty/files/11674533/Types.Scala.3.pdf)

### CODE REVIEW

File: docs/_spec/03-types.md [hunk 1]
```diff
@@ -672,73 +1009,93 @@ We define the following relations between types.
...
  [^argisnotwildcard]: In these cases, if `T_i` and/or `U_i` are wildcard type arguments, the [simplification rules](#simplification-rules) for parameterized types allow to reduce them to real types.
 
-A declaration or definition in some compound type of class type ´C´ _subsumes_ another declaration of the same name in some compound type or class type ´C'´, if one of the following holds.
-
-- A value declaration or definition that defines a name ´x´ with type ´T´ subsumes a value or method declaration that defines ´x´ with type ´T'´, provided ´T <: T'´.
-- A method declaration or definition that defines a name ´x´ with type ´T´ subsumes a method declaration that defines ´x´ with type ´T'´, provided ´T <: T'´.
-- A type alias `type ´t´[´T_1´, ..., ´T_n´] = ´T´` subsumes a type alias `type ´t´[´T_1´, ..., ´T_n´] = ´T'´` if ´T \equiv T'´.
-- A type declaration `type ´t´[´T_1´, ..., ´T_n´] >: ´L´ <: ´U´` subsumes a type declaration `type ´t´[´T_1´, ..., ´T_n´] >: ´L'´ <: ´U'´` if ´L' <: L´ and ´U <: U'´.
-- A type or class definition that binds a type name ´t´ subsumes an abstract type declaration `type t[´T_1´, ..., ´T_n´] >: L <: U` if ´L <: t <: U´.
-
```
[the remark this answers is not in the document]
Comment (odersky) [reply docs/_spec/03-types.md:741]: The section on Volatile types should be dropped. It is replaced by a notion of realizability.

The section on Erasure should be updated to the new rules. 

Generally, all usages of compound types (using `with`) should be dropped.

**Review by odersky — CHANGES_REQUESTED** [review]
I like the new structure. It's a great improvement to what we had before.

**Review by odersky — CHANGES_REQUESTED** [review]
I like the new structure. It's a great improvement to what we had before.

File: docs/_spec/03-types.md [hunk 2]
```diff
@@ -7,185 +7,455 @@ chapter: 3
...
+2. Replace every implicit or explicit reference to `this` in the refinement declarations by ´\alpha´.
+3. Create nested [refined types](#refined-types), one for every refined declaration.
+4. Unless ´\alpha´ was never actually used, wrap the result in a [recursive type](#recursive-types) `{ ´\alpha´ => ´...´ }`.
 
-### Type Projection
+### Concrete Type Lambdas
 
 ```ebnf
-SimpleType  ::=  SimpleType ‘#’ id
+TypeLambda            ::= TypeLambdaParams ‘=>>’ Type
```
[the remark this answers is not in the document]
Comment (sjrd) [reply docs/_spec/03-types.md:234]: The official rendering is HTML. I posted the PDF as an easy-to-distribute form, but I directly printed from the browser to get it. If we need the PDF to be a truly public document, it seems there are a number of other things that don't render well; notably lines seem to be cut across pages.

File: docs/_spec/03-types.md [hunk 3]
```diff
@@ -7,185 +7,455 @@ chapter: 3
...
+A _parameterized type_ ´T[T_1, ..., T_n]´ consists of a type constructor ´T´ and type arguments ´T_1, ..., T_n´ where ´n \geq 1´.
+The parameterized type is well-formed if
+
+- ´T´ is a type constructor which takes ´n´ type parameters ´a_1, ..., a_n´, i.e., it must conform to a type lambda of the form ´[\pm a_1 >: L_1 <: H_1, ..., \pm a_n >: L_n <: H_n] => U´, and
+- if ´T´ is an abstract type constructor, none of the type arguments is a wildcard type argument, and
+- each type argument _conforms to its bounds_, i.e., given ´\sigma´ the substitution ´[a_1 := T_1, ..., a_n := T_n]´, for each type ´i´:
+  - if ´T_i´ is a type and ´\sigma L_i <: T_i <: \sigma H_i´, or
+  - ´T_i´ is a wildcard type argument ´? >: L_{Ti} <: H_{Ti}´ and ´\sigma L_i <: L_{Ti}´ and ´H_{Ti} <: \sigma H_i´.
+
+´T[T_1, ..., T_n]´ is an _applied class type_ if and only if ´T´ is a [class type](#type-designators).
```
[the remark this answers is not in the document]
Comment (sjrd) [reply docs/_spec/03-types.md:431]: Ah yes, parameterized class type is better.

File: docs/_spec/03-types.md [hunk 4]
```diff
@@ -425,49 +731,54 @@ class B extends C[B] with D with E
...
+Skolem types are exclusively used temporarily during subtyping derivations.
+They are stable types.
 
-Function types are shorthands for class types that define an `apply` method.
-Specifically, the ´n´-ary function type ´(T_1, ..., T_n) \Rightarrow R´ is a shorthand for the class type `Function´_n´[´T_1´, ..., ´T_n´, ´R´]`.
-In particular ´() \Rightarrow R´ is a shorthand for class type `Function´_0´[´R´]`.
+A skolem type of the form ´∃ \alpha : T´ represents a stable reference to unknown value of type ´T´.
+The identifier ´\alpha´ is chosen uniquely every time a skolem type is created.
+However, as a skolem type is stable, it can be substituted in several occurrences in other types.
+When "copied" through substutition, all the copies retain the same ´\alpha´, and are therefore equivalent.
```
[the remark this answers is not in the document]
Comment (sjrd) [reply docs/_spec/03-types.md:748]: I feel like that is what the paragraph just above says. Is there something missing or unclear there?

> Skolem types cannot directly be written in the concrete syntax. Moreover, although they are proper types, they are never inferred as the type given to terms. Skolem types are exclusively used temporarily during subtyping derivations.

[the remark this answers is not in the document]
Comment (odersky) [reply docs/_spec/03-types.md:748]: Where does it say that skolem types are not part of inferred types?

[the remark this answers is not in the document]
Comment (sjrd) [reply docs/_spec/03-types.md:748]: > they are never inferred as the type given to terms

I guess you're saying we should say they are never inferred *anywhere*, including type arguments and things like that?

[the remark this answers is not in the document]
Comment (odersky) [reply docs/_spec/03-types.md:748]: No, the rule is that they cannot be part of inferred types of vals or defs.

File: docs/_spec/03-types.md [hunk 5]
```diff
@@ -303,83 +571,121 @@ The following type adds the `@suspendable` annotation to the type `String`:
...
 
 ### Union and Intersection Types
 
-Syntactically, the types `S | T` and `S & T` are infix types, where the infix operators are `|` and `&`, respectively (see above).
+```ebnf
+UnionType         ::=  Type ‘｜‘ Type
+IntersectionType  ::=  Type ‘＆‘ Type
+```
+
+Syntactically, the types `S | T` and `S & T` are infix types, where the infix operators are `|` and `&`, respectively (see [infix types](#infix-types)).
```
[the remark this answers is not in the document]
Comment (sjrd) [reply docs/_spec/03-types.md:688]: > In line 711 below we have: "If ´C´ is invariant, emit a compile error". What is that supposed to mean? I can certainly write such a type:
> 
> ```scala
> class Ref[X]
> // defined class Ref
>                                                                                                                        
> scala> def x: Ref[Int] & Ref[String] = ???
> ```

Hum, this came from the [Scala 3 reference](https://docs.scala-lang.org/scala3/reference/new-types/intersection-types-spec.html). But I agree that it doesn't seem to make sense. I'll remove the third rule and only keep the first two.

File: docs/_spec/03-types.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (sjrd) [reply docs/_spec/03-types.md:741]: Weak conformance will be the subject of a separate PR that introduces harmonization, so I would rather not touch it now.

I updated erasure. For the volatile types I'll have to investigate how realizability affects that section.

**Comment by sjrd** [comment]
I think I have addressed everything so far except the Volatile Types -> Realizable comment.

**Comment by sjrd** [comment]
Volatility now replaced by realizability as well.

**Pull request merged**